src/main/java/com/firebolt/jdbc/type/BaseType.java

```diff
@@ -19,16 +18,20 @@
 import java.sql.Timestamp;
 import java.util.Arrays;
 import java.util.TimeZone;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
 
+import static com.firebolt.jdbc.exception.ExceptionType.TYPE_TRANSFORMATION_ERROR;
 import static com.firebolt.jdbc.type.array.SqlArrayUtil.BYTE_ARRAY_PREFIX;
 import static com.firebolt.jdbc.type.array.SqlArrayUtil.hexStringToByteArray;
 
 /** This class contains the java types the Firebolt datatypes are mapped to */
 public enum BaseType {
-	LONG(Long.class, conversion -> Long.parseLong(checkInfinity(conversion.getValue()))),
-	INTEGER(Integer.class, conversion -> Integer.parseInt(checkInfinity(conversion.getValue()))),
-	SHORT(Short.class, conversion -> Short.parseShort(checkInfinity(conversion.getValue()))),
-	BIGINT(BigInteger.class, conversion -> new BigInteger(checkInfinity(conversion.getValue()))),
+	LONG(TypePredicate.mayBeFloatingNumber, Long.class, conversion -> Long.parseLong(checkInfinity(conversion.getValue())), conversion -> Double.valueOf(conversion.getValue()).longValue()),
+	INTEGER(TypePredicate.mayBeFloatingNumber, Integer.class, conversion -> Integer.parseInt(checkInfinity(conversion.getValue())), conversion -> Double.valueOf(conversion.getValue()).intValue()),
+	SHORT(TypePredicate.mayBeFloatingNumber, Short.class, conversion -> Short.parseShort(checkInfinity(conversion.getValue())), conversion -> Double.valueOf(conversion.getValue()).shortValue()),
+	BYTE(TypePredicate.mayBeFloatingNumber, Byte.class, conversion -> Byte.parseByte(checkInfinity(conversion.getValue())), conversion -> Double.valueOf(conversion.getValue()).byteValue()),
```

Comment:
Why do we have `Double.valueOf` in the second function in each type instead of the actual type?

Reply:
The second function can convert floating point number to specific integer type (byte, short, int, long). So it has to parse the number as floating point (either Float or Double) and then take it integer part only. Double is preferable here because it can work with bigger number and it is the java default. Type Float was introduced into java 30 years ago when computers had 16M RAM and 1G disk.

Reply:
Ok, makes sense.